moshi/src/test/java/com/squareup/moshi/TypesTest.java

```diff
@@ -321,6 +321,40 @@ public static Type getFirstTypeArgument(Type type) throws Exception {
     }
   }
 
+  //
+  // Regression tests for https://github.com/square/moshi/issues/338
+  //
+  // Adapted from https://github.com/google/gson/pull/1128
+  //
+
+  private static final class RecursiveTypeVars<T> {
+    RecursiveTypeVars<? super T> superType;
+  }
+
+  @Test public void recursiveTypeVariablesResolve() {
+    JsonAdapter<RecursiveTypeVars<String>> adapter = new Moshi.Builder().build().adapter(Types
+        .newParameterizedTypeWithOwner(TypesTest.class, RecursiveTypeVars.class, String.class));
+    assertThat(adapter).isNotNull();
+  }
+
+  @Test public void testRecursiveTypeVariablesResolve1() {
+    JsonAdapter<TestType> adapter = new Moshi.Builder().build().adapter(TestType.class);
+    assertThat(adapter).isNotNull();
+  }
+
+  @Test public void testRecursiveTypeVariablesResolve12() {
```

Comment:
This is the nittiest of nits, but might as well fix gson's typo here (should be "Resolve2", not "Resolve12").

Reply:
5136b6e